# Pandas Industry-Based Hands-On Exercises — Complete Solution

# HANDS-ON EXERCISE 1 — E-Commerce Customer Analysis

**Industry:** E-commerce / Retail

## Part A — Load the Data

The task says to import Pandas, load the dataset, investigate a delimiter problem if necessary, correct it, and display the first five records.

The supplied `customer_sales.csv` is already comma-separated. We verify the structure instead of assuming a delimiter problem exists.


In [ ]:
# First try the normal CSV delimiter.
sales_raw = pd.read_csv("customer_sales.csv")

print("Shape:", sales_raw.shape)
print("Number of columns:", len(sales_raw.columns))
print("Columns:", sales_raw.columns.tolist())
sales_raw.head()


Shape: (5000, 7)
Number of columns: 7
Columns: ['Customer ID', 'Customer Name', 'City', 'Age', 'Product Category', 'Sales', 'Order Status']


,Customer ID,Customer Name,City,Age,Product Category,Sales,Order Status
0,CUST00001,Joseph Obi,Lagos,65,Beauty,"61,241.84",Completed
1,CUST00002,Tolu Onyeka,Abuja,55,Home Appliances,"7,969.88",Completed
2,CUST00003,Fatima Ogunleye,Ilorin,19,Accessories,"49,729.44",Pending
3,CUST00004,Mary Akinyemi,Ibadan,46,Office Supplies,"69,565.72",Returned
4,CUST00005,Kelechi Abubakar,Owerri,45,Sports,"69,485.89",Completed


### Delimiter investigation

The file opens into **7 separate columns**, so the supplied file does not currently exhibit the unusual-delimiter problem described in the scenario. If it had loaded as one column, we would have inspected the raw separator and reloaded it with the correct `sep=` argument.

For this supplied file, using the normal comma delimiter is the correct result.


In [ ]:
# Final DataFrame for Exercise 1
df_sales = pd.read_csv("customer_sales.csv")

print("First 5 records:")
display(df_sales.head())


First 5 records:


,Customer ID,Customer Name,City,Age,Product Category,Sales,Order Status
0,CUST00001,Joseph Obi,Lagos,65,Beauty,"61,241.84",Completed
1,CUST00002,Tolu Onyeka,Abuja,55,Home Appliances,"7,969.88",Completed
2,CUST00003,Fatima Ogunleye,Ilorin,19,Accessories,"49,729.44",Pending
3,CUST00004,Mary Akinyemi,Ibadan,46,Office Supplies,"69,565.72",Returned
4,CUST00005,Kelechi Abubakar,Owerri,45,Sports,"69,485.89",Completed


## Part B — Understand the Dataset

The required inspection uses `head()`, `shape`, `info()`, and `describe()`.


In [ ]:
print("Number of rows:", df_sales.shape[0])
print("Number of columns:", df_sales.shape[1])

print("\nColumn names:")
print(df_sales.columns.tolist())

print("\nData types and non-null counts:")
df_sales.info()

print("\nBasic numerical statistics:")
display(df_sales.describe())

Number of rows: 5000
Number of columns: 7

Column names:
['Customer ID', 'Customer Name', 'City', 'Age', 'Product Category', 'Sales', 'Order Status']

Data types and non-null counts:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Customer ID       5000 non-null   object 
 1   Customer Name     5000 non-null   object 
 2   City              5000 non-null   object 
 3   Age               5000 non-null   int64  
 4   Product Category  5000 non-null   object 
 5   Sales             5000 non-null   float64
 6   Order Status      5000 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 273.6+ KB

Basic numerical statistics:


,Age,Sales
count,"5,000.00","5,000.00"
mean,43.91,"124,393.02"
std,15.36,"71,795.87"
min,18.00,37.88
25%,31.00,"61,591.57"
50%,44.00,"124,713.01"
75%,57.00,"186,804.63"
max,70.00,"249,810.49"


### Part C — Clean the Data

The exercise asks us to investigate:

- Missing City values
- Text where numerical values are expected
- Unnecessary rows

In the supplied sales dataset, there are **no missing values**, the numerical columns are already numeric, and there are no exact duplicate rows. Therefore, there is no unnecessary deletion to perform.

This is an important analyst decision: **do not change clean data just to demonstrate a cleaning operation.**


In [ ]:
print("Missing values by column:")
display(df_sales.isna().sum())

print("Rows containing at least one missing value:", df_sales.isna().any(axis=1).sum())
print("Exact duplicate rows:", df_sales.duplicated().sum())

# Confirm the numerical columns are actually numeric.
print("\nSales dtype:", df_sales["Sales"].dtype)
print("Age dtype:", df_sales["Age"].dtype)


Missing values by column:


Customer ID         0
Customer Name       0
City                0
Age                 0
Product Category    0
Sales               0
Order Status        0
dtype: int64

Rows containing at least one missing value: 0
Exact duplicate rows: 0

Sales dtype: float64
Age dtype: int64


## Part D — Analyze

### 1. How many customers are in each city?


In [ ]:
customers_by_city = df_sales["City"].value_counts()
display(customers_by_city)


City
Enugu            364
Abuja            360
Calabar          358
Warri            348
Uyo              346
Kaduna           338
Kano             337
Port Harcourt    336
Ilorin           331
Abeokuta         327
Benin City       325
Jos              321
Lagos            308
Owerri           303
Ibadan           298
Name: count, dtype: int64

### 2. What is the average sales value?


In [ ]:
average_sales = df_sales["Sales"].mean()
print(f"Average sales value: {average_sales:,.2f}")


Average sales value: 124,393.02


### 3. Which product category appears most frequently?


In [ ]:
category_counts = df_sales["Product Category"].value_counts()
display(category_counts)
print("Most frequent product category:", category_counts.idxmax())


Product Category
Home Appliances    549
Office Supplies    524
Accessories        507
Books              505
Beauty             499
Clothing           497
Sports             484
Electronics        483
Furniture          478
Groceries          474
Name: count, dtype: int64

Most frequent product category: Home Appliances


### 4. How many completed orders are there?


In [ ]:
completed_orders = (df_sales["Order Status"] == "Completed").sum()
print("Completed orders:", completed_orders)


Completed orders: 3099


### 5. How many records contain missing information?


In [ ]:
missing_sales_records = df_sales.isna().any(axis=1).sum()
print("Records containing missing information:", missing_sales_records)


Records containing missing information: 0


### Exercise 1 interpretation

The supplied e-commerce dataset is already structurally clean: there are no missing records or exact duplicates. The average sales value is about **124,393.02**, and **Home Appliances** is the most frequent product category. There are **3,099 completed orders**.

The city counts show the distribution of customer records across the cities; these counts can be used for basic geographic customer analysis.


## Exercise 1 — Challenge

Create `Customer_Value` using:

- Sales >= 100,000 → **High Value**
- Sales < 100,000 → **Standard**

Then determine how many High Value customers are in the dataset.


In [ ]:
df_sales["Customer_Value"] = df_sales["Sales"].apply(
    lambda x: "High Value" if x >= 100000 else "Standard"
)

display(df_sales[["Customer ID", "Sales", "Customer_Value"]].head(10))

high_value_count = (df_sales["Customer_Value"] == "High Value").sum()
print("High Value customers:", high_value_count)

print("\nCustomer value distribution:")
display(df_sales["Customer_Value"].value_counts())


,Customer ID,Sales,Customer_Value
0,CUST00001,"61,241.84",Standard
1,CUST00002,"7,969.88",Standard
2,CUST00003,"49,729.44",Standard
3,CUST00004,"69,565.72",Standard
4,CUST00005,"69,485.89",Standard
5,CUST00006,"24,201.68",Standard
6,CUST00007,"134,068.62",High Value
7,CUST00008,"221,365.80",High Value
8,CUST00009,"193,272.76",High Value
9,CUST00010,"69,511.45",Standard


High Value customers: 2983

Customer value distribution:


Customer_Value
High Value    2983
Standard      2017
Name: count, dtype: int64

### Challenge interpretation

There are **2,983 High Value customers** and **2,017 Standard customers**. Under the exercise's rule, most records fall into the High Value category. This is a classification based only on the supplied sales threshold; it should not be interpreted as a complete measure of customer profitability.


# HANDS-ON EXERCISE 2 — Hospital Patient Data Cleaning

**Industry:** Healthcare

## Part A — Initial Inspection

The task requires `head()`, `shape`, `info()`, and `describe()`, followed by documentation of what we discover.


In [ ]:
df_hospital_raw = pd.read_csv("hospital_patient_data_cleaning_practice.csv")

display(df_hospital_raw.head())

print("Shape:", df_hospital_raw.shape)

print("\nInfo:")
df_hospital_raw.info()

print("\nDescriptive statistics:")
display(df_hospital_raw.describe(include="all"))


,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
0,1745,Sani Olawale,58.00,Male,Asaba,3,2026,Specialist Consultation
1,1679,Daniel James,49.00,Male,Owerri,10,2025,Specialist Consultation
2,3291,John Ibekwe,20.00,Female,Asaba,9,2026,Routine Check
3,3489,Bola Sani,56.00,Female,Ilorin,2,2024,Routine Check
4,1943,Aisha Chukwu,50.00,Male,Lagos,11,2026,Routine Check


Shape: (3000, 8)

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Patient_ID    3000 non-null   int64  
 1   Patient_Name  3000 non-null   object 
 2   Age           2955 non-null   float64
 3   Gender        3000 non-null   object 
 4   City          2881 non-null   object 
 5   Visit_Count   2959 non-null   object 
 6   Record_Year   3000 non-null   int64  
 7   Visit_Type    3000 non-null   object 
dtypes: float64(1), int64(2), object(5)
memory usage: 187.6+ KB

Descriptive statistics:


,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
count,"3,000.00",3000,"2,955.00",3000,2881,2959,"3,000.00",3000
unique,NaN,1320,NaN,3,23,17,NaN,5
top,NaN,Daniel Etim,NaN,Female,Calabar,1,NaN,Routine Check
freq,NaN,9,NaN,1505,178,256,NaN,1120
mean,"2,511.95",NaN,49.01,NaN,NaN,NaN,"2,025.24",NaN
std,905.75,NaN,18.14,NaN,NaN,NaN,0.76,NaN
min,"1,001.00",NaN,0.00,NaN,NaN,NaN,"2,024.00",NaN
25%,"1,751.00",NaN,34.00,NaN,NaN,NaN,"2,025.00",NaN
50%,"2,504.50",NaN,49.00,NaN,NaN,NaN,"2,025.00",NaN
75%,"3,255.25",NaN,65.00,NaN,NaN,NaN,"2,026.00",NaN


### What was discovered

The supplied hospital dataset contains:

- Missing City values.
- Missing Age values.
- Missing Visit_Count values.
- Text values in `Visit_Count`, including values such as `seven`, `ten`, `six`, and `4 visits`.
- Exact duplicate rows.
- Inconsistent city labels such as `ABUJA`, `ibadan `, `lagos`, and `Port-Harcourt`.
- Records clearly marked as `TEST` with dummy/test patient information.

The exercise explicitly says to investigate before deleting problematic data, so the next cells quantify these issues first.


## Part B — Investigate Data Problems


In [ ]:
print("Missing values by column:")
display(df_hospital_raw.isna().sum())

print("Exact duplicate rows:", df_hospital_raw.duplicated().sum())

print("\nUnexpected Visit_Count values:")
visit_numeric_test = pd.to_numeric(df_hospital_raw["Visit_Count"], errors="coerce")
display(df_hospital_raw.loc[
    visit_numeric_test.isna() & df_hospital_raw["Visit_Count"].notna(),
    ["Patient_ID", "Patient_Name", "Visit_Count"]
].drop_duplicates().head(30))

print("\nPotential test/dummy records:")
display(df_hospital_raw[
    df_hospital_raw["City"].fillna("").astype(str).str.strip().str.upper().eq("TEST")
])

print("\nCity values that need normalization:")
display(df_hospital_raw["City"].value_counts(dropna=False).sort_index())


Missing values by column:


Patient_ID        0
Patient_Name      0
Age              45
Gender            0
City            119
Visit_Count      41
Record_Year       0
Visit_Type        0
dtype: int64

Exact duplicate rows: 4

Unexpected Visit_Count values:


,Patient_ID,Patient_Name,Visit_Count
6,1290,Sarah Olawale,six
46,2673,Helen Williams,4 visits
96,1999,Anthony Olawale,seven
143,2136,Yusuf Yusuf,ten
153,1638,Ruth Mensah,six
197,3859,Daniel Onyeka,seven
301,1464,Hauwa Adeyemi,ten
363,2565,Victor Musa,seven
387,3816,David Mohammed,4 visits
401,2189,Ruth Cole,ten



Potential test/dummy records:


,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
18,9005,System Test,0.00,Unknown,TEST,NaN,2024,Test
64,9001,Test Patient One,0.00,Unknown,TEST,NaN,2026,Test
355,9004,Dummy Patient,1.00,Unknown,TEST,1,2025,Test
1399,9003,Clinic Test Entry,0.00,Unknown,TEST,NaN,2026,Test
2461,9002,Sample Record,99.00,Unknown,TEST,0,2026,Test



City values that need normalization:


City
ABUJA             15
Abeokuta         154
Abuja            144
Akure            150
Asaba            164
Benin City       158
Calabar          178
Enugu            165
Ibadan           128
Ilorin           161
Jos              140
Kaduna           160
Kano             165
Lagos            147
Owerri           139
Port Harcourt    149
Port-Harcourt     15
TEST               5
Uyo              174
Warri            170
Zaria            170
ibadan            15
lagos             15
NaN              119
Name: count, dtype: int64

## Cleaning decisions

1. **Exact duplicate rows:** remove them because they contain the same complete record.
2. **Test/dummy records:** remove rows whose City is `TEST`; these are explicitly labelled as test/sample records rather than genuine patient activity.
3. **City:** replace missing values with `Unknown`, as required by the exercise.
4. **Inconsistent city spelling/capitalization:** normalize labels, including `ABUJA` → `Abuja`, `ibadan ` → `Ibadan`, `lagos` → `Lagos`, and `Port-Harcourt` → `Port Harcourt`.
5. **Visit_Count:** convert recoverable text values (`seven`, `ten`, `six`, `4 visits`) into numbers. Rows still missing Visit_Count are removed because visit count is required for patient classification and average visit calculations.
6. **Age:** missing Age values are retained because Age is not required for any of the requested calculations. Filling them with an invented value would add an unsupported assumption.


In [ ]:
df_hospital = df_hospital_raw.drop_duplicates().copy()

# Remove clearly labelled test/dummy records.
is_test_record = df_hospital["City"].fillna("").astype(str).str.strip().str.upper().eq("TEST")
test_records_removed = is_test_record.sum()
df_hospital = df_hospital.loc[~is_test_record].copy()

# Fill missing cities as required.
df_hospital["City"] = df_hospital["City"].fillna("Unknown")

# Normalize inconsistent city labels.
df_hospital["City"] = df_hospital["City"].astype(str).str.strip().str.title()
df_hospital["City"] = df_hospital["City"].replace({
    "Port-Harcourt": "Port Harcourt"
})

# Clean Visit_Count.
visit_clean = (
    df_hospital["Visit_Count"]
    .astype(str)
    .str.strip()
    .str.lower()
    .replace({"seven": "7", "ten": "10", "six": "6"})
    .str.replace(" visits", "", regex=False)
)

df_hospital["Visit_Count"] = pd.to_numeric(visit_clean, errors="coerce")

# Rows with no usable visit count cannot be classified or included in visit averages.
invalid_visit_rows = df_hospital["Visit_Count"].isna().sum()
df_hospital = df_hospital.dropna(subset=["Visit_Count"]).copy()

print("Test/dummy records removed:", test_records_removed)
print("Rows with unusable Visit_Count removed:", invalid_visit_rows)
print("Records remaining after cleaning:", len(df_hospital))
print("\nRemaining missing values:")
display(df_hospital.isna().sum())


Test/dummy records removed: 5
Rows with unusable Visit_Count removed: 38
Records remaining after cleaning: 2953

Remaining missing values:


Patient_ID       0
Patient_Name     0
Age             44
Gender           0
City             0
Visit_Count      0
Record_Year      0
Visit_Type       0
dtype: int64

## Part C — Missing Values

The exercise specifically requires missing City values to become `Unknown`. The cleaning above applies and verifies that rule.


In [ ]:
print("Unknown City records after cleaning:", (df_hospital["City"] == "Unknown").sum())
print("Missing City values after cleaning:", df_hospital["City"].isna().sum())


Unknown City records after cleaning: 118
Missing City values after cleaning: 0


## Part D — Patient Classification

Create `Patient_Category`:

- Visit count >= 5 → **Frequent Patient**
- Visit count < 5 → **Occasional Patient**


In [ ]:
df_hospital["Patient_Category"] = df_hospital["Visit_Count"].apply(
    lambda x: "Frequent Patient" if x >= 5 else "Occasional Patient"
)

display(df_hospital[["Patient_ID", "Visit_Count", "Patient_Category"]].head(10))


,Patient_ID,Visit_Count,Patient_Category
0,1745,3.00,Occasional Patient
1,1679,10.00,Frequent Patient
2,3291,9.00,Frequent Patient
3,3489,2.00,Occasional Patient
4,1943,11.00,Frequent Patient
5,3956,7.00,Frequent Patient
6,1290,6.00,Frequent Patient
7,1434,1.00,Occasional Patient
8,3830,6.00,Frequent Patient
9,3475,5.00,Frequent Patient


## Part E — Quick Analysis

### 1. Number of patients by city


In [ ]:
patients_by_city = df_hospital["City"].value_counts()
display(patients_by_city)


City
Calabar          174
Uyo              170
Zaria            169
Warri            167
Enugu            163
Asaba            163
Port Harcourt    162
Kano             161
Lagos            161
Ilorin           159
Kaduna           158
Benin City       155
Abuja            155
Abeokuta         151
Akure            147
Ibadan           143
Owerri           139
Jos              138
Unknown          118
Name: count, dtype: int64

### 2. Number of frequent patients


In [ ]:
frequent_patients = (df_hospital["Patient_Category"] == "Frequent Patient").sum()
print("Frequent patients:", frequent_patients)


Frequent patients: 1986


### 3. Average patient visit count


In [ ]:
average_visits = df_hospital["Visit_Count"].mean()
print(f"Average patient visit count: {average_visits:.2f}")


Average patient visit count: 6.52


### 4. Number of records with missing information before cleaning


In [ ]:
hospital_missing_records_before = df_hospital_raw.isna().any(axis=1).sum()
print("Records with at least one missing value before cleaning:", hospital_missing_records_before)


Records with at least one missing value before cleaning: 204


### 5. Number of records remaining after cleaning


In [ ]:
print("Records remaining after cleaning:", len(df_hospital))


Records remaining after cleaning: 2953


### Hospital analysis interpretation

After the documented cleaning steps, **2,953 records remain**. There are **1,986 Frequent Patients**, and the average visit count is approximately **6.52 visits**.

The city distribution should be interpreted after normalization, because otherwise capitalization and spelling differences would make the same city appear as separate groups.


## Exercise 2 — Challenge

Create a summary table with:

| Patient Category | Number of Patients | Average Visits |
|---|---:|---:|
| Frequent Patient | ? | ? |
| Occasional Patient | ? | ? |


In [ ]:
hospital_summary = (
    df_hospital.groupby("Patient_Category")["Visit_Count"]
    .agg(Number_of_Patients="count", Average_Visits="mean")
    .reindex(["Frequent Patient", "Occasional Patient"])
)

display(hospital_summary)


,Number_of_Patients,Average_Visits
Patient_Category,,
Frequent Patient,1986,8.48
Occasional Patient,967,2.51


### Challenge interpretation

The Frequent Patient group contains **1,986 patients** and averages about **8.48 visits**, while the Occasional Patient group contains **967 patients** and averages about **2.51 visits**. This shows a clear difference in visit activity between the two groups and gives the hospital a useful starting point for investigating why some patients return much more often.


# HANDS-ON EXERCISE 3 — Banking Customer Transaction Analysis

**Industry:** Banking / Financial Services

## Part A — Load and Inspect


In [ ]:
df_banking_raw = pd.read_csv("banking_customer_transaction.csv")

display(df_banking_raw.head())

print("Dataset size:", df_banking_raw.shape)

print("\nColumn names:")
print(df_banking_raw.columns.tolist())

print("\nData types:")
df_banking_raw.info()

print("\nNumerical summary statistics:")
display(df_banking_raw.describe(include="all"))


,Customer ID,Customer Name,City,Transaction Type,Amount,Account Status
0,CUST00585,Tunde Okafor,Kaduna,Online Transfer,37974.75,Active
1,CUST01061,Bola Uche,Lagos,Deposit,9896.89,Active
2,CUST00077,Kelechi Nwosu,Abuja,Transfer,59302.75,Active
3,CUST00898,Tunde Abubakar,Benin City,Deposit,71719.97,Dormant
4,CUST00594,Amina Ojo,Lagos,Online Transfer,3979.55,Active


Dataset size: (3000, 6)

Column names:
['Customer ID', 'Customer Name', 'City', 'Transaction Type', 'Amount', 'Account Status']

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Customer ID       3000 non-null   object
 1   Customer Name     2990 non-null   object
 2   City              2976 non-null   object
 3   Transaction Type  2982 non-null   object
 4   Amount            2884 non-null   object
 5   Account Status    2988 non-null   object
dtypes: object(6)
memory usage: 140.8+ KB

Numerical summary statistics:


,Customer ID,Customer Name,City,Transaction Type,Amount,Account Status
count,3000,2990,2976,2982,2884,2988
unique,1031,400,10,7,2803,4
top,CUST00532,Femi Ekanem,Lagos,Transfer,Not Available,Active
freq,9,16,731,548,20,2371


## Part B — Investigate the Transaction Amount

We inspect missing values, unexpected text, and invalid records before cleaning.


In [ ]:
print("Missing values by column:")
display(df_banking_raw.isna().sum())

amount_numeric_test = pd.to_numeric(df_banking_raw["Amount"], errors="coerce")

print("\nUnexpected/non-numeric Amount values:")
display(df_banking_raw.loc[
    amount_numeric_test.isna() & df_banking_raw["Amount"].notna(),
    ["Customer ID", "Amount"]
].drop_duplicates().head(50))

print("\nExact duplicate rows:", df_banking_raw.duplicated().sum())


Missing values by column:


Customer ID           0
Customer Name        10
City                 24
Transaction Type     18
Amount              116
Account Status       12
dtype: int64


Unexpected/non-numeric Amount values:


,Customer ID,Amount
11,CUST00372,Error
51,CUST00092,Not Available
155,CUST00945,Error
182,CUST01039,Error
184,CUST00061,Pending
281,CUST00772,Not Available
284,CUST01073,Pending
322,CUST00146,Error
363,CUST00299,Pending
374,CUST00302,Unknown



Exact duplicate rows: 0


### Cleaning decisions

- Missing categorical values are represented as `Unknown` so the records are not unnecessarily discarded.
- Currency-formatted values such as `₦12,500` can be converted safely to numeric.
- `1O000` is treated as a recoverable typing error where the letter `O` is clearly being used in place of zero, so it is converted to `10000`.
- Placeholder/error values such as `Not Available`, `Error`, `Pending`, `Unknown`, and `##VALUE!` cannot be reliably inferred as transaction amounts, so their Amount becomes missing and those rows are removed from amount-based analysis.
- Repeated Customer IDs are **not** automatically removed because a banking customer can legitimately have multiple transactions. There are no exact duplicate rows.


In [ ]:
df_banking = df_banking_raw.copy()

# Fill missing categorical information without inventing a category.
for col in ["Customer Name", "City", "Transaction Type", "Account Status"]:
    df_banking[col] = df_banking[col].fillna("Unknown")

# Clean recoverable numeric formatting/typos.
amount_clean = (
    df_banking["Amount"]
    .astype(str)
    .str.replace("₦", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.replace("O", "0", regex=False)
)

df_banking["Amount"] = pd.to_numeric(amount_clean, errors="coerce")

invalid_amount_rows = df_banking["Amount"].isna().sum()
df_banking = df_banking.dropna(subset=["Amount"]).copy()

print("Rows with unusable transaction amounts removed:", invalid_amount_rows)
print("Records remaining after cleaning:", len(df_banking))
print("\nRemaining missing values:")
display(df_banking.isna().sum())


Rows with unusable transaction amounts removed: 186
Records remaining after cleaning: 2814

Remaining missing values:


Customer ID         0
Customer Name       0
City                0
Transaction Type    0
Amount              0
Account Status      0
dtype: int64

## Part C — Clean the Dataset

The cleaning above is intentionally documented: recoverable formatting errors are repaired, while values that cannot safely be reconstructed are excluded from calculations that require a numeric transaction amount.


## Part D — Create Transaction Classification

- Amount >= 100,000 → **Large Transaction**
- Amount < 100,000 → **Regular Transaction**


In [ ]:
df_banking["Transaction_Category"] = df_banking["Amount"].apply(
    lambda x: "Large Transaction" if x >= 100000 else "Regular Transaction"
)

display(df_banking[["Customer ID", "Amount", "Transaction_Category"]].head(10))


,Customer ID,Amount,Transaction_Category
0,CUST00585,"37,974.75",Regular Transaction
1,CUST01061,"9,896.89",Regular Transaction
2,CUST00077,"59,302.75",Regular Transaction
3,CUST00898,"71,719.97",Regular Transaction
4,CUST00594,"3,979.55",Regular Transaction
5,CUST00796,"7,614.28",Regular Transaction
6,CUST00107,"31,818.41",Regular Transaction
7,CUST00957,"20,408.71",Regular Transaction
8,CUST00013,"27,533.50",Regular Transaction
9,CUST00998,"11,931.24",Regular Transaction


## Part E — Business Analysis

### 1. How many transactions occurred in each city?


In [ ]:
transactions_by_city = df_banking["City"].value_counts()
display(transactions_by_city)


City
Lagos            686
Abuja            466
Port Harcourt    332
Ibadan           291
Kano             286
Enugu            181
Kaduna           175
Benin City       149
Aba              121
Jos              104
Unknown           23
Name: count, dtype: int64

### 2. What is the average transaction amount?


In [ ]:
average_transaction = df_banking["Amount"].mean()
print(f"Average transaction amount: {average_transaction:,.2f}")


Average transaction amount: 45,076.66


### 3. How many large transactions occurred?


In [ ]:
large_transactions = (df_banking["Transaction_Category"] == "Large Transaction").sum()
print("Large transactions:", large_transactions)


Large transactions: 276


### 4. Which transaction type appears most frequently?


In [ ]:
transaction_type_counts = df_banking["Transaction Type"].value_counts()
display(transaction_type_counts)
print("Most frequent transaction type:", transaction_type_counts.idxmax())


Transaction Type
Transfer           511
Deposit            468
Withdrawal         438
Payment            377
ATM Withdrawal     354
POS Purchase       354
Online Transfer    295
Unknown             17
Name: count, dtype: int64

Most frequent transaction type: Transfer


### 5. What is the average transaction amount by transaction type?


In [ ]:
average_by_transaction_type = (
    df_banking.groupby("Transaction Type")["Amount"]
    .mean()
    .sort_values(ascending=False)
)
display(average_by_transaction_type)


Transaction Type
POS Purchase      50,549.25
Withdrawal        48,961.16
Online Transfer   46,682.39
Transfer          46,319.72
Unknown           46,222.93
Deposit           44,458.18
Payment           40,820.96
ATM Withdrawal    36,960.13
Name: Amount, dtype: float64

### Banking analysis interpretation

After cleaning, **2,814 transaction records** remain for amount-based analysis. The average transaction amount is approximately **45,076.66**, and there are **276 Large Transactions**.

`Transfer` is the most frequent transaction type in the cleaned data. The grouped averages show how typical transaction values differ across transaction types.


## Exercise 3 — Challenge

Use `groupby()` to determine the average transaction amount by city. Then identify the city with the highest average and explain what it could mean from a business-analysis perspective.


In [ ]:
average_by_city = (
    df_banking.groupby("City")["Amount"]
    .mean()
    .sort_values(ascending=False)
)

display(average_by_city)

highest_city = average_by_city.idxmax()
highest_city_average = average_by_city.max()

print(f"City with the highest average transaction amount: {highest_city}")
print(f"Average transaction amount in {highest_city}: {highest_city_average:,.2f}")


City
Benin City      52,154.37
Aba             50,332.42
Lagos           46,905.36
Kaduna          46,313.14
Abuja           46,284.95
Jos             44,224.25
Kano            43,659.66
Ibadan          42,323.28
Port Harcourt   41,283.75
Enugu           40,047.12
Unknown         33,783.73
Name: Amount, dtype: float64

City with the highest average transaction amount: Benin City
Average transaction amount in Benin City: 52,154.37


### Challenge interpretation

**Benin City** has the highest average transaction amount at approximately **52,154.37**. This could indicate that the transactions represented in the dataset from Benin City tend to be larger on average.

However, this is an **observation, not a final business decision**. The bank should investigate transaction volume, customer mix, account types, transaction types, and the distribution of unusually large transactions before concluding why the city has a higher average.


# HANDS-ON EXERCISE 4 — Energy Consumption Analysis

**Industry:** Energy / Utilities

## Part A — Understand the Data


In [ ]:
df_energy_raw = pd.read_csv("energy_consumption.csv")

display(df_energy_raw.head())

print("Number of records:", df_energy_raw.shape[0])
print("Number of columns:", df_energy_raw.shape[1])

print("\nData types:")
df_energy_raw.info()

print("\nSummary statistics:")
display(df_energy_raw.describe(include="all"))


,Customer ID,City,Customer Type,Monthly Consumption,Account Status,Energy Source
0,EN00001,Abuja,Residential,111.5,Active,Grid Electricity
1,EN00002,Lagos,Residential,104.0,Active,Grid Electricity
2,EN00003,Lagos,Residential,671.5,Active,Grid Electricity
3,EN00004,Abuja,Residential,172.6,Active,Grid Electricity
4,EN00005,Port Harcourt,Small Business,440.9,Active,Grid + Solar


Number of records: 3012
Number of columns: 6

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3012 entries, 0 to 3011
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Customer ID          3012 non-null   object
 1   City                 2970 non-null   object
 2   Customer Type        2994 non-null   object
 3   Monthly Consumption  2894 non-null   object
 4   Account Status       2997 non-null   object
 5   Energy Source        2992 non-null   object
dtypes: object(6)
memory usage: 141.3+ KB

Summary statistics:


,Customer ID,City,Customer Type,Monthly Consumption,Account Status,Energy Source
count,3012,2970,2994,2894,2997,2992
unique,3000,10,5,2508,4,5
top,EN00100,Lagos,Residential,Pending,Active,Grid Electricity
freq,2,643,1548,21,2430,1661


## Part B — Data Cleaning

We investigate:

- Missing City values
- Missing consumption values
- Unexpected text values
- Unnecessary rows

The dataset contains exact duplicate rows, missing categorical values, and non-numeric consumption placeholders. Some malformed numeric strings are recoverable, such as `12O` and `5,2O0`.


In [ ]:
print("Missing values by column:")
display(df_energy_raw.isna().sum())

print("Exact duplicate rows:", df_energy_raw.duplicated().sum())

consumption_numeric_test = pd.to_numeric(df_energy_raw["Monthly Consumption"], errors="coerce")

print("\nUnexpected/non-numeric consumption values:")
display(df_energy_raw.loc[
    consumption_numeric_test.isna() & df_energy_raw["Monthly Consumption"].notna(),
    ["Customer ID", "Monthly Consumption"]
].drop_duplicates().head(50))


Missing values by column:


Customer ID              0
City                    42
Customer Type           18
Monthly Consumption    118
Account Status          15
Energy Source           20
dtype: int64

Exact duplicate rows: 12

Unexpected/non-numeric consumption values:


,Customer ID,Monthly Consumption
8,EN00009,Unknown
82,EN00083,No Reading
85,EN00086,Error
91,EN00092,Unknown
107,EN00108,No Reading
151,EN00152,Unknown
163,EN00164,No Reading
165,EN00166,12O
183,EN00184,Error
191,EN00192,"5,2O0"


### Energy cleaning decisions

1. Remove exact duplicate rows.
2. Replace missing categorical information with `Unknown`.
3. Correct clearly recoverable numeric formatting such as commas and the letter `O` used in place of zero.
4. Convert remaining non-numeric placeholders to missing.
5. Remove records with no usable Monthly Consumption because the requested consumption averages and categories cannot be calculated for them.


In [ ]:
df_energy = df_energy_raw.drop_duplicates().copy()

for col in ["City", "Customer Type", "Account Status", "Energy Source"]:
    df_energy[col] = df_energy[col].fillna("Unknown")

consumption_clean = (
    df_energy["Monthly Consumption"]
    .astype(str)
    .str.replace(",", "", regex=False)
    .str.replace("O", "0", regex=False)
)

df_energy["Monthly Consumption"] = pd.to_numeric(consumption_clean, errors="coerce")

invalid_consumption_rows = df_energy["Monthly Consumption"].isna().sum()
df_energy = df_energy.dropna(subset=["Monthly Consumption"]).copy()

print("Rows with unusable consumption removed:", invalid_consumption_rows)
print("Records remaining after cleaning:", len(df_energy))
print("\nRemaining missing values:")
display(df_energy.isna().sum())


Rows with unusable consumption removed: 187
Records remaining after cleaning: 2813

Remaining missing values:


Customer ID            0
City                   0
Customer Type          0
Monthly Consumption    0
Account Status         0
Energy Source          0
dtype: int64

## Part C — Create Consumption Category

- Consumption >= 500 → **High Consumption**
- Consumption < 500 → **Normal Consumption**


In [ ]:
df_energy["Consumption_Category"] = df_energy["Monthly Consumption"].apply(
    lambda x: "High Consumption" if x >= 500 else "Normal Consumption"
)

display(df_energy[["Customer ID", "Monthly Consumption", "Consumption_Category"]].head(10))


,Customer ID,Monthly Consumption,Consumption_Category
0,EN00001,111.50,Normal Consumption
1,EN00002,104.00,Normal Consumption
2,EN00003,671.50,High Consumption
3,EN00004,172.60,Normal Consumption
4,EN00005,440.90,Normal Consumption
5,EN00006,"3,113.60",High Consumption
6,EN00007,403.60,Normal Consumption
7,EN00008,206.80,Normal Consumption
9,EN00010,213.40,Normal Consumption
10,EN00011,421.00,Normal Consumption


## Part D — Analyze Consumption

### 1. Number of customers in each city


In [ ]:
customers_by_energy_city = df_energy["City"].value_counts()
display(customers_by_energy_city)


City
Lagos            604
Abuja            428
Port Harcourt    349
Ibadan           301
Kano             286
Enugu            215
Benin City       175
Kaduna           159
Aba              139
Jos              120
Unknown           37
Name: count, dtype: int64

### 2. Average consumption


In [ ]:
average_consumption = df_energy["Monthly Consumption"].mean()
print(f"Average monthly consumption: {average_consumption:,.2f}")


Average monthly consumption: 1,078.73


### 3. Number of high-consumption customers


In [ ]:
high_consumption_count = (df_energy["Consumption_Category"] == "High Consumption").sum()
print("High-consumption customers:", high_consumption_count)


High-consumption customers: 1284


### 4. Average consumption by customer type


In [ ]:
average_consumption_by_type = (
    df_energy.groupby("Customer Type")["Monthly Consumption"]
    .mean()
    .sort_values(ascending=False)
)
display(average_consumption_by_type)


Customer Type
Industrial           6,464.33
Commercial           1,784.30
Public Institution   1,060.19
Unknown                964.65
Small Business         736.89
Residential            306.01
Name: Monthly Consumption, dtype: float64

### 5. Number of customers using each energy source


In [ ]:
energy_source_counts = df_energy["Energy Source"].value_counts()
display(energy_source_counts)


Energy Source
Grid Electricity    1544
Grid + Solar         435
Solar                310
Generator            306
Grid + Generator     199
Unknown               19
Name: count, dtype: int64

### Energy analysis interpretation

After cleaning, **2,813 records** remain. Average monthly consumption is approximately **1,078.73**, and **1,284 customers** fall into the High Consumption category.

Industrial customers have the highest average consumption in the cleaned data, while Residential customers have the largest number of records. These are useful patterns for further investigation, but they are not by themselves a final business decision.


## Exercise 4 — Challenge

Create a summary showing:

- Customer Type
- Average Consumption
- Number of Customers

Then write a short recommendation for what the energy company should investigate next.


In [ ]:
energy_summary = (
    df_energy.groupby("Customer Type")["Monthly Consumption"]
    .agg(
        Average_Consumption="mean",
        Number_of_Customers="count"
    )
    .sort_values("Average_Consumption", ascending=False)
)

display(energy_summary)


,Average_Consumption,Number_of_Customers
Customer Type,,
Industrial,"6,464.33",194
Commercial,"1,784.30",409
Public Institution,"1,060.19",133
Unknown,964.65,17
Small Business,736.89,610
Residential,306.01,1450


### Challenge recommendation

The energy company should investigate **why Industrial customers have a much higher average monthly consumption than the other customer types**, including whether this is explained by customer size, operating hours, equipment, tariff structure, or the number of high-consumption accounts.

The company should also examine the unusually high-consumption records individually before making operational or pricing decisions. This follows the exercise's instruction to identify a useful pattern for further investigation rather than making a final business decision from a simple analysis.


# Key Final Results

| Exercise | Key result |
|---|---|
| E-Commerce | Average sales ≈ **124,393.02**; completed orders = **3,099**; High Value customers = **2,983** |
| Hospital | Records remaining = **2,953**; Frequent Patients = **1,986**; average visits ≈ **6.52** |
| Banking | Records remaining = **2,814**; average transaction ≈ **45,076.66**; Large Transactions = **276**; highest city average = **Benin City** |
| Energy | Records remaining = **2,813**; average consumption ≈ **1,078.73**; High Consumption = **1,284**; Industrial has the highest average by customer type |
